Block/Service/AbstractCategoriesBlockService.php

```diff
@@ -0,0 +1,188 @@
+
+        return $this->renderResponse($blockContext->getTemplate(), array(
+            'context' => $blockContext,
+            'settings' => $blockContext->getSettings(),
+            'block' => $blockContext->getBlock(),
+            'category' => $category,
+            'root' => $root,
+        ), $response);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildEditForm(FormMapper $formMapper, BlockInterface $block)
+    {
+        $adminField = $this->getFormAdminType($formMapper, $this->categoryAdmin, 'categoryId', 'category', array(
+            'label' => 'form.label_category',
+        ), array(
+            'translation_domain' => 'SonataClassificationBundle',
+            'link_parameters' => array(
+                array(
+                    array(
+                        'context' => $block->getSetting('context'),
+                    ),
+                ),
+            ),
+        ));
+
+        $formMapper->add('settings', 'sonata_type_immutable_array', array(
+            'keys' => array(
+                array('title', 'text', array(
+                    'label' => 'form.label_title',
+                    'required' => false,
+                )),
+                array('context', 'choice', array(
+                    'label' => 'form.label_context',
+                    'required' => false,
+                    'choices' => $this->getContextChoices(),
+                )),
+                array($adminField, null, array()),
+            ),
+            'translation_domain' => 'SonataClassificationBundle',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureSettings(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults(array(
+            'title' => 'Categories',
+            'category' => false,
+            'categoryId' => null,
+            'context' => 'default',
+            'template' => 'SonataClassificationBundle:Block:base_block_categories.html.twig',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function load(BlockInterface $block)
+    {
+        if (is_numeric($block->getSetting('categoryId'))) {
+            $block->setSetting('categoryId', $this->getCategory($block->getSetting('categoryId')));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function prePersist(BlockInterface $block)
+    {
+        $this->resolveIds($block);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function preUpdate(BlockInterface $block)
+    {
+        $this->resolveIds($block);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getBlockMetadata($code = null)
+    {
+        $description = (!is_null($code) ? $code : $this->getName());
+
+        return new Metadata($this->getName(), $description, false, 'SonataClassificationBundle', array(
+            'class' => 'fa fa-folder-open-o',
+        ));
+    }
+
+    /**
+     * @param mixed $id
```

Comment:
Based on the code this can be integer and some kind of object? Can we narrow it down? Is there a base category or interface? (I think there is) It would make easier to read what this method does.